bouncer/src/repo/manipulator/modeloptimizer/repo_optimizer_multipart.cpp

```diff
@@ -84,207 +86,559 @@ bool MultipartOptimizer::collectMeshData(
+		}
+
+		if (mapped.uvChannels.size() == 0 && submUVs.size() != 0)
+		{
+			//initialise uvChannels
+			mapped.uvChannels.resize(submUVs.size());
+		}
+
+		if (mapped.uvChannels.size() == submUVs.size())
+		{
+			for (uint32_t i = 0; i < submUVs.size(); ++i)
+			{
+				mapped.uvChannels[i].insert(mapped.uvChannels[i].end(), submUVs[i].begin(), submUVs[i].end());
+			}
+		}
+		else
+		{
+			//This shouldn't happen, if it does, then it means the mFormat isn't set correctly
+			repoError << "Unexpected transformedMesh format mismatch occured!";
 		}
 	}
 	else
 	{
-		repoError << "Scene or node is null!";
+		repoError << "Failed merging meshes: Vertices or faces cannot be null!";
 	}
+}
 
-	return success;
+void updateMinMax(repo::lib::RepoVector3D& min, repo::lib::RepoVector3D& max, repo::lib::RepoVector3D v)
+{
+	if (v.x < min.x) {
+		min.x = v.x;
+	}
+	if (v.y < min.y) {
+		min.y = v.y;
+	}
+	if (v.z < min.z) {
+		min.z = v.z;
+	}
+
+	if (v.x > max.x) {
+		max.x = v.x;
+	}
+	if (v.y > max.y) {
+		max.y = v.y;
+	}
+	if (v.z > max.z) {
+		max.z = v.z;
+	}
 }
 
-repo::core::model::MeshNode* MultipartOptimizer::createSuperMesh
-(
-	const repo::core::model::RepoScene *scene,
-	const std::set<repo::lib::RepoUUID>           &meshGroup,
-	std::unordered_map<repo::lib::RepoUUID, repo::lib::RepoUUID, repo::lib::RepoUUIDHasher>  &matIDs,
-	const bool isGrouped)
+void MultipartOptimizer::splitMeshes(
+	const repo::core::model::RepoScene* scene,
+	repo::core::model::MeshNode node,
+	std::vector<mapped_mesh_t>& mappedMeshes
+)
 {
-	std::vector<repo::lib::RepoVector3D> vertices, normals;
-	std::vector<repo_face_t> faces;
-	std::vector<std::vector<repo::lib::RepoVector2D>> uvChannels;
-	std::vector<repo_color4d_t> colors;
-	std::vector<repo_mesh_mapping_t> meshMapping;
+	auto start = boost::chrono::high_resolution_clock::now();
 
-	repo::core::model::MeshNode* resultMesh = nullptr;
+	using Scalar = float;
+	using Bvh = bvh::Bvh<Scalar>;
+	using Vector3 = bvh::Vector3<Scalar>;
+
+	// Create a set of bounding boxes & centers for each Face in the oversized 
+	// mesh.
+	// The BVH builder expects a set of bounding boxes and centers to work with.
+
+	auto faces = node.getFaces();
+	auto vertices = node.getVertices();
+	auto boundingBoxes = std::vector<bvh::BoundingBox<Scalar>>();
+
+	for (auto& face : faces)
+	{
+		auto v = vertices[face[0]];
+		auto b = bvh::BoundingBox<Scalar>(Vector3(v.x, v.y, v.z));
 
-	std::vector<repo::core::model::MeshNode*> resultMeshes;
-	repo::lib::RepoMatrix startMat;
+		for (int i = 1; i < face.size(); i++)
+		{
+			v = vertices[face[i]];
+			b.extend(Vector3(v.x, v.y, v.z));
+		}
 
-	bool success = collectMeshData(scene, scene->getRoot(defaultGraph), meshGroup, startMat,
-		vertices, normals, faces, uvChannels, colors, meshMapping, matIDs);
+		boundingBoxes.push_back(b);
+	}
 
-	if (success && meshMapping.size())
+	auto centers = std::vector<Vector3>();
+	for (auto& bounds : boundingBoxes)
```

Comment:
Updated! 
(I don't like this part of the Bvh API because it feels like we have to maintain two computations, when one is clearly a dependent of the other. Why build() can't just work it out when center() is on the type I don't know! (or at least give a compute_bounding_boxes_centers to go with compute_bounding_boxes_union)